src/routes/credentials.rs

```diff
@@ -281,23 +295,74 @@ async fn post_credential_handler(
         .get(&oidc_context.client_id)
         .ok_or(Error::OauthInvalidClientId)?;
     if let Some(checks_directory) = &client_config.checks_directory {
-        let checker = app_state
+        let mut app_state_write = app_state.write().await;
+        let checker = app_state_write
             .rhai_checkers
             .get_or_create(&oidc_context.client_id, checks_directory)?;
         checker.check(&id_token)?;
+        drop(app_state_write);
     }
 
-    // return the response as a HTTP NoContent, to give the frontend a chance to do the redirect on its own
-    Ok(HttpResponse::NoContent()
-        .append_header((
-            "Location",
-            format!(
-                "{}#id_token={}&refresh_token={}&state={}&token_type=bearer",
-                oidc_context.redirect_uri.clone(),
-                id_token,
-                refresh_token,
-                oidc_context.state.clone(),
-            ),
-        ))
-        .finish())
+    let response_type = session
+        .get::<String>(RESPONSE_TYPE_SESSION_KEY)?
+        .ok_or(Error::ResponseType)?;
+
+    drop(app_state_read);
+
+    if response_type == "code" {
+        log::info!("Authorization Code Flow");
+        // Generate a random URL encoded string as the Authorization Code.
+        let mut bytes = vec![0; 45];
+        let mut rng = rand::thread_rng();
+        rng.fill(&mut bytes[..]);
+        let code = base64::engine::general_purpose::URL_SAFE.encode(bytes);
+
+        // Store (code -> token_response) so it can be sent later at the `/token` endpoint.
+        let token_response = TokenResponse {
+            token_type: "Bearer".to_string(),
+            access_token: "null".to_string(),
+            refresh_token,
+            id_token,
+        };
+        token_storage
+            .insert(
+                code.clone(),
+                (
+                    token_response.clone(),
+                    session
+                        .get(REDIRECT_URI_SESSION_KEY)?
+                        .ok_or(Error::RedirectUri)?,
+                ),
+            )
+            .await;
+
+        // return the response as a HTTP NoContent, to give the frontend a chance to do the redirect on its own.
+        Ok(HttpResponse::NoContent()
+            .append_header((
+                "Location",
+                format!(
+                    "{}?code={}&state={}",
+                    oidc_context.redirect_uri.clone(),
+                    code,
+                    oidc_context.state.clone(),
+                ),
+            ))
+            .finish())
+    } else if response_type == "id_token" {
+        log::info!("Implicit flow");
+        Ok(HttpResponse::NoContent()
+            .append_header((
+                "Location",
+                format!(
+                    "{}#id_token={}&refresh_token={}&state={}&token_type=bearer",
```

Comment:
Should bearer start with capital B here as well?

Reply:
No, it shouldn't at least from my knowledge.

Reply:
From what I saw. If it references the token type it should be lowercase.